### DISCUSSION TITLE
Future of the opencollective-api layer - NestJS

### DISCUSSION THREAD

**Discussion by SudharakaP** [opencollective/opencollective#4044] [Ideas]
Hi everyone, 😄 

This goes out to @opencollective/core team. And I hope this this isn't too over the top/overambitious kinda suggestion and forgive me if it is. 😸 😃 

Something that I felt for some time as an external contributor is that our API layer is a bit convoluted. The code is a bit harder to find in my own humble opinion and experience. Maybe it's just me but I normally have to debug stuff to find where things are coming from. 😄 

Also one thing I noticed for some time is that the contributions (external contributions especially) to the API is far less than the front-end side (maybe this is due to more front-end related changes in the projects). In any case I felt that maybe having a structured framework approach (such as provided by NextJS on the front end side) will help the API layer too and wondering if there's any plans to migrate or rewrite the API in something like [NestJS](https://nestjs.com/)? I feel this might help scale, as the API grows in size, and probably also provide a consistent architecture to the API that developers have to follow (at least to some extent). I am not a NestJS expert btw, but if this is of interest, I am quite interested and willing to help on this any way I can. 😄 😄

**Comment by Betree** [comment]
Most of the complexity in the API is there because of historical reasons. For example, the fact that we currently have both GraphQL V1 and V2 is a transitional state, we'll ultimately have only one. Emails are also supposed to move to their own service when we'll have the bandwidth to do so.

I think people contributing more to the frontend than API is a normal thing, I've seen a similar trend on other projects. I guess there are more frontend developers, and it's probably simpler to discover a project by working on the frontend.

I'm not sure about the benefits that a framework like NestJS could bring in our case. It surely helps to design and structure large projects, but there are a lot of conventions already in place in the existing codebase and even though the structure is not perfect, it's there and it's working fine. Moving to a new framework has a high engineering cost, both to make the migration and to deal with the learning curve. There must be strong and clearly identified benefits before making such a move.

In the short/mid-term re-organizing folder structures, improving the docs, and adding some helpers/libs to make the developer experience better is something we're open to.

**Reply by SudharakaP** [reply]
> Most of the complexity in the API is there because of historical reasons. For example, the fact that we currently have both GraphQL V1 and V2 is a transitional state, we'll ultimately have only one. Emails are also supposed to move to their own service when we'll have the bandwidth to do so.
> 
> I think people contributing more to the frontend than API is a normal thing, I've seen a similar trend on other projects. I guess there are more frontend developers, and it's probably simpler to discover a project by working on the frontend.
> 
> I'm not sure about the benefits that a framework like NestJS could bring in our case. It surely helps to design and structure large projects, but there are a lot of conventions already in place in the existing codebase and even though the structure is not perfect, it's there and it's working fine. Moving to a new framework has a high engineering cost, both to make the migration and to deal with the learning curve. There must be strong and clearly identified benefits before making such a move.

Thanks for the detailed reply and clarification. Makes perfect sense.  😄  

> In the short/mid-term re-organizing folder structures, improving the docs, and adding some helpers/libs to make the developer experience better is something we're open to.

Let me think how we can do some improvements in this regard. 🤔
